embedded/sql/stmt.go

```diff
@@ -439,109 +509,146 @@ func (stmt *UpsertIntoStmt) validate(table *Table) (map[uint64]int, error) {
 		selByColID[col.id] = i
 	}
 
-	if !pkIncluded {
+	if !pkIncluded && (!stmt.isInsert || !table.pk.autoIncrement) {
 		return nil, ErrPKCanNotBeNull
 	}
 
+	if pkIncluded && stmt.isInsert && table.pk.autoIncrement {
+		return nil, ErrNoValueForAutoIncrementalColumn
+	}
+
 	return selByColID, nil
 }
 
-func (stmt *UpsertIntoStmt) compileUsing(e *Engine, implicitDB *Database, params map[string]interface{}) (ces, des []*store.KV, db *Database, err error) {
+func (stmt *UpsertIntoStmt) compileUsing(e *Engine, implicitDB *Database, params map[string]interface{}) (summary *TxSummary, err error) {
+	summary = newTxSummary()
+	summary.db = implicitDB
+
+	if implicitDB == nil {
+		return nil, ErrNoDatabaseSelected
+	}
+
 	table, err := stmt.tableRef.referencedTable(e, implicitDB)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, err
 	}
 
 	cs, err := stmt.validate(table)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, err
 	}
 
 	for _, row := range stmt.rows {
 		if len(row.Values) != len(stmt.cols) {
-			return nil, nil, nil, ErrInvalidNumberOfValues
+			return nil, ErrInvalidNumberOfValues
 		}
 
-		pkVal := row.Values[cs[table.pk.id]]
+		cols := stmt.cols
+
+		var pkVal ValueExp
+
+		// inject auto-incremental pk value
+		if stmt.isInsert && table.pk.autoIncrement {
+			table.maxPK++
+			e.catalog.mutated = true // TODO: implement transactional in-memory catalog
```

Comment:
Why is catalog reload needed here?

Reply:
because the greatest key so far it's kept inside the in-mem representation of the catalog. But it's possible to distinguish which part of the catalog needs to be reloaded.
I think this catalog implementation deserves an analysis, which may be addressed as a separate work

Reply:
OK so please correct me if I'm wrong - this is to revert the `maxPK` to an original value by reloading the catalog if there's any error and the transaction doesn't really commit?